doc/user/methods/nmf.md

```diff
@@ -0,0 +1,624 @@
+arma::mat W, H;
+
+mlpack::NMF nmf;                         // Step 1: create object.
+double residue = nmf.Apply(V, 10, W, H); // Step 2: apply NMF to decompose V.
+
+// Now print some information about the factorized matrices.
+std::cout << "W has size: " << W.n_rows << " x " << W.n_cols << "."
+    << std::endl;
+std::cout << "H has size: " << H.n_rows << " x " << H.n_cols << "."
+    << std::endl;
+std::cout << "RMSE of reconstructed matrix: "
+    << arma::norm(V - W * H, "fro") / std::sqrt(V.n_elem) << "." << std::endl;
+```
+<p style="text-align: center; font-size: 85%"><a href="#simple-examples">More examples...</a></p>
+
+#### Quick links:
+
+ * [Constructors](#constructors): create `NMF` objects.
+ * [`Apply()`](#applying-decompositions): apply `NMF` decomposition to data.
+ * [Examples](#simple-examples) of simple usage and links to detailed example
+   projects.
+ * [Template parameters](#advanced-functionality-template-parameters) for
+   using different update rules, initialization strategies, and termination
+   criteria.
+ * [Advanced template examples](#advanced-functionality-examples) of use with
+   custom template parameters.
+
+#### See also:
+
+<!-- TODO: add these links
+ * [`AMF`](amf.md): alternating matrix factorization
+ * [`CF`](cf.md): collaborative filtering (recommender system)
+ * [`SparseCoding`](sparse_coding.md)
+-->
+
+ * [mlpack transformations](../../index.md#transformations)
+ * [Non-negative matrix factorization on Wikipedia](https://en.wikipedia.org/wiki/Non-negative_matrix_factorization)
+ * [Learning the parts of objects by non-negative matrix factorization](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=29bae9472203546847ec1352a604566d0f602728) (original NMF paper, pdf)
+
+### Constructors
+
+ * `nmf = NMF()`
+   - Create an `NMF` object.
+   - The rank of the decomposition is specified in the call to
+     [`Apply()`](#applying-decompositions).
+
+---
+
+ * `nmf = NMF(SimpleResidueTermination(minResidue=1e-5, maxIterations=10000))`
+   - Create an NMF object with custom termination parameters.
+   - `minResidue` (a `double`) specifies the minimum difference of the norm of
+     `W * H` between iterations for termination.
+   - `maxIterations` specifies the maximum number of iterations before
+     decomposition terminates.
+
+---
+
+### Applying Decompositions
+
+ * `double residue = nmf.Apply(V, rank, W, H)`
+   - Decompose the matrix `V` into two non-negative matrices `W` and `H` with
+     rank `rank`.
+   - `W` will be set to size `V.n_rows` x `rank`.
+   - `H` will be set to size `rank` x `V.n_cols`.
+   - `W` and `H` are initialized randomly using the
+     [Acol](#advanced-functionality-initializationruletype) initialization
+     strategy; i.e., each column of `W` is an average of 5 random columns of
+     `V`, and `H` is initialized uniformly randomly.
+   - The residue (change in the norm of `W * H` between iterations) is returned.
+
```

Comment:
This is really nice, I was able to understand really well NMF, even I did not know anything about it before you have briefly explained it to me in Amsterdam last year.
The only one thing that is missing in my opinion is to add a range for the `rank` values (e.g., what is a typical low and high values for rank).
Also if there are typical values or range for `residue`, it would be really nice to add that as well.
Apart from that, everything looks good.
:+1: :100:

Reply:
I did some amount of surveying the literature and this is actually a complicated topic: there is no easy way to know what rank is best, and a lot of the heuristic techniques boil down to "try it a lot and see what looks right according to some loss measure", which... isn't really great.  So there is not really much to suggest as a good default.  I tried to clarify a little bit about the tradeoff in f707299982.